### PR TITLE
Set production variant name

### LINKED ISSUE #6178: [FR] Currently unable to specify variant names when deploying models to Sagemaker
### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

I would like to add the ability to pass in a specific name for the model variant when deploying to Sagmaker as an additional argument to the deploy method in sagemaker/__init__.py.  Currently, variant names are autogenerated by _get_sagemaker_model_name using the endpoint name, model name and uid.  

### Motivation

> #### What is the use case for this feature?
To supply a specific name to each model variant hosted on Sagemaker.

> #### Why is this use case valuable to support for MLflow users in general?
The name would be more meaningful than the current default.

> #### Why is this use case valuable to support for your project(s) or organization?
This would be especially important for endpoints that host multiple models or different version of the same model (ex: for testing purposes).  It would make it much easier for the user to better understand the purpose of a given model based on its name and distinguish between models.  

> #### Why is it currently difficult to achieve this use case?
Currently, there is not way to specify the variant name when deploying, they are just generated automatically

### Details

I propose adding an argument to the deploy method where a specific name can be entered.  This will throw an error if that name currently exists or if it violates any naming conventions.  Use the current default method if no name is specified when calling this method.

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [X] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [X] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by nfarley-soaren**
## Related Issues/PRs 

Close #6178
## What changes are proposed in this pull request?

The option to manually specify a production variant name rather than using the default model name when deploying a model to an endpoint

## How is this patch tested?

- [x] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.
- I tested variant name creation, as well as the ability to error if using improper names (won't match sagemaker regex), or trying to create a variant with the same name as an existing variant.  The code used for testing requires access to a custom docker image rather than the default image created by mlflow so it is not attached.

 I tested variant name creation, as well as the ability to error if using improper names (won't match sagemaker regex), or trying to create a variant with the same name as an existing variant.  The code used for testing requires access to a custom docker image rather than the default image created by mlflow so it is not attached.

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x ] Yes. Make sure the changed pages / sections render correctly by following the steps below.
If I knew where this was, I'd indicate the pages.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)
Allows user to specify a name for a given production variant when deploying mlflow model to Sagemaker endpoint.  If no name is given, the default is used.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/sagemaker/__init__.py [hunk 1]
```diff
@@ -163,6 +164,7 @@ def push_image_to_ecr(image=DEFAULT_IMAGE_NAME):
 def deploy(
     app_name,
     model_uri,
+    variant_name=None,
```
Comment (dbczumar) [inline mlflow/sagemaker/__init__.py:167]: Nit: For backwards compatibility, can we add `variant_name` to the end of the argument list?

File: mlflow/sagemaker/__init__.py [hunk 2]
```diff
@@ -388,6 +394,17 @@ def deploy(
...
+    if variant_name:
+        pattern = r"[a-zA-Z0-9](-*[a-zA-Z0-9]){0,62}"
+        if not re.fullmatch(pattern, variant_name):
+            raise MlflowException(
+            message=(
+                "Variant name must match the following pattern" 
+                " {variant_pattern}".format(variant_pattern=pattern)
+            ),
+            error_code=INVALID_PARAMETER_VALUE,
+        )
```
Comment (dbczumar) [inline mlflow/sagemaker/__init__.py:406]: Do we need to validate the variant name in the client, or can SageMaker perform the validation?

File: mlflow/sagemaker/__init__.py [hunk 3]
```diff
@@ -1591,15 +1616,26 @@ def _update_sagemaker_endpoint(
...
     if mode == DEPLOYMENT_MODE_ADD:
         new_model_weight = 0
         production_variants = deployed_production_variants
+        variant_name_list = [variant["VariantName"] for variant in production_variants]
+        if variant_name and variant_name in variant_name_list:
+            raise MlflowException(
+            message=(
+                "Variant {variant_name} already exists!".format(variant_name=variant_name)
+            ),
+            error_code=INVALID_PARAMETER_VALUE,
```
Comment (dbczumar) [inline mlflow/sagemaker/__init__.py:1631]: Does SageMaker enforce variant name uniqueness? If not, does the client strictly need to perform this validation?

**Review by dbczumar — COMMENTED** [review]
@nfarley-soaren Apologies for the delay. Awesome work! I left a few small comments / questions. Before merging, can we also add support for this argument to https://mlflow.org/docs/latest/python_api/mlflow.sagemaker.html#mlflow.sagemaker.SageMakerDeploymentClient.create_deployment and https://mlflow.org/docs/latest/python_api/mlflow.sagemaker.html#mlflow.sagemaker.SageMakerDeploymentClient.update_deployment? This can be accomplished by forwarding the argument to `deploy()` here: https://github.com/mlflow/mlflow/blob/904608bbd35907cee15bddaba7396b68e4989221/mlflow/sagemaker/__init__.py#L2133 and here: https://github.com/mlflow/mlflow/blob/904608bbd35907cee15bddaba7396b68e4989221/mlflow/sagemaker/__init__.py#L2368.

Can we also update the docs for `mlflow.sagemaker.deploy()` and the `SageMakerDeploymentClient` APIs to include small examples of `production_variant` usage?

File: mlflow/sagemaker/__init__.py (hunk 2 above)
Comment (nfarley-soaren) [reply mlflow/sagemaker/__init__.py:406]: We don't have to.  Sagemaker will throw an error out if it is invalid.  I can remove this if that is preferred

File: mlflow/sagemaker/__init__.py (hunk 1 above)
Comment (nfarley-soaren) [reply mlflow/sagemaker/__init__.py:167]: Can do.

File: mlflow/sagemaker/__init__.py (hunk 2 above)
Comment (dbczumar) [reply mlflow/sagemaker/__init__.py:406]: Awesome! Yeah, I think we can remove this :)

File: mlflow/sagemaker/__init__.py (hunk 3 above)
Comment (nfarley-soaren) [reply mlflow/sagemaker/__init__.py:1631]: I will look into this.

Comment (dbczumar) [reply mlflow/sagemaker/__init__.py:1617]: Thanks @nfarley-soaren ! Looking forward to merging once you finish your investigation.

File: mlflow/sagemaker/__init__.py [hunk 4]
```diff
@@ -1563,7 +1573,8 @@ def _update_sagemaker_endpoint(
     :param role: SageMaker execution ARN role.
     :param sage_client: A boto3 client for SageMaker.
     :param s3_client: A boto3 client for S3.
-    """
+    :variant_name: (optional) The name to assign to the new production variant if it doesn't already exist.
+                     - If variant_name already exists an MlflowException will be raised.    """
     if mode not in [DEPLOYMENT_MODE_ADD, DEPLOYMENT_MODE_REPLACE]:
```
Comment (dbczumar) [inline mlflow/sagemaker/__init__.py:1578]: Nit:

```suggestion
                     - If variant_name already exists an MlflowException will be raised.
    """
    if mode not in [DEPLOYMENT_MODE_ADD, DEPLOYMENT_MODE_REPLACE]:
```

File: mlflow/sagemaker/__init__.py [hunk 5]
```diff
@@ -200,7 +202,7 @@ def deploy(
                       For more information about supported URI schemes, see
                       `Referencing Artifacts <https://www.mlflow.org/docs/latest/concepts.html#
                       artifact-locations>`_.
-
+
```
Comment (dbczumar) [inline mlflow/sagemaker/__init__.py:205]:
```suggestion

```

Comment (dbczumar) [inline mlflow/sagemaker/__init__.py:205]:
```suggestion

```

File: mlflow/sagemaker/__init__.py [hunk 6]
```diff
@@ -2,6 +2,7 @@
 The ``mlflow.sagemaker`` module provides an API for deploying MLflow models to Amazon SageMaker.
 """
 import os
+import re
```
Comment (dbczumar) [inline mlflow/sagemaker/__init__.py:5]: Nit: I think we can remove this unused import

File: mlflow/sagemaker/__init__.py (hunk 3 above)
Comment (nfarley-soaren) [reply mlflow/sagemaker/__init__.py:1617]: Thank you @dbczumar.  Looks like Sagemaker will error out if the variant name already exists so I'll remove this part also.

**Comment by nfarley-soaren** [comment]
I'm not sure where to update the docs with examples.

**Review by dbczumar — APPROVED** [review]
@nfarley-soaren I pushed some small docs tweaks to your branch and cleaned up the formatting of some existing docs inline examples. This is looking great! LGTM & thank you so much for your contribution!

**Pull request merged**